### PR TITLE
AP_Compass_QMC5883P.cpp: Adding QMC5883P Compass into the system still…

### PR DESCRIPTION
**Pull request by Lokesh-Carbonix**
As we already have a QMC5883L here I am trying to add QMC5883P.

Currently the probe method is not being called and also tried to print using GCS_SEND_TEXT() but didn't worked.

The code is able to compile for AP_Periph but not working as a features.

### CODE REVIEW

File: libraries/AP_Compass/AP_Compass.h [hunk 1]
```diff
@@ -433,6 +433,7 @@ friend class AP_Compass_Backend;
         DRIVER_MSP      =17,
         DRIVER_SERIAL   =18,
         DRIVER_MMC5XX3  =19,
+		DRIVER_QMC5883P =20,
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass.h:436]: try to match whitespace with previous lines

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) 2016  Emlid Ltd. All rights reserved.
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:2]: why copyright emlid? better to remove the line completely and rely on git history

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+//Reset
+#define QMC5883P_RST 0x80
+
+extern const AP_HAL::HAL &hal;
+
+AP_Compass_Backend *AP_Compass_QMC5883P::probe(AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                                               bool force_external,
+                                               enum Rotation rotation)
+{
+	GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "Before Probe");
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: remove these debug lines

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	if (!dev) {
+        return nullptr;
+    }
+	GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "dev found");
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: remove

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    AP_Compass_QMC5883P *sensor = new AP_Compass_QMC5883P(std::move(dev),force_external,rotation);
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+	GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "sensor found");
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: remove

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+#if 0
+    _dump_registers();
+#endif
+    GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "Before Whoami");
+    if(!_check_whoami()){
+         GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "Before Whoami Fail");
+    	 goto fail;
+    }
+
+    if (!_dev->write_register(0x0B, 0x01)||
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: try to get whitespace right
for new files, use Tools/CodeStyle/ardupilot-astyle.sh

Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: add a define name for registers? or an enum?

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+ fail:
+    _dev->get_semaphore()->give();
+    return false;
+}
+bool AP_Compass_QMC5883P::_check_whoami()
+{
+    uint8_t whoami;
+    //Affected by other devices,must read registers 0x00 once or reset,after can read the ID registers reliably
+    _dev->read_registers(0x00,&whoami,1);
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: register names

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    		whoami != QMC5883P_ID_VAL){
+    	return false;
+    }
+    return true;
+}
+
+void AP_Compass_QMC5883P::timer()
+{
+    struct PACKED {
+    	le16_t rx;
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:171]: stm32 and other ardupilot MCUs are little endian already, so just use int16_t and avoid the casts

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    //new data is ready
+    if (!(status & 0x01)) {
+    	return;
+    }
+
+  if(!_dev->read_registers(QMC5883P_REG_DATA_OUTPUT_X, (uint8_t *) &buffer, sizeof(buffer))){
+	  return ;
+  }
+
+    auto x = -static_cast<int16_t>(le16toh(buffer.rx));
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:191]: need to explain the sign changes, or is this really a rotation?
better to use "normal" orientation for sensor, then in hwdef.dat setup rotation. Less annoying for the next vendor who uses a different rotation

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    printf("mag.x:%d\n",x);
+    printf("mag.y:%d\n",y);
+    printf("mag.z:%d\n",z);
+#endif
+
+    Vector3f field = Vector3f{x * range_scale , y * range_scale, z * range_scale };
+
+    // rotate to the desired orientation
+    if (is_external(_instance)) {
+        field.rotate(ROTATION_YAW_90);
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:205]: why yaw_90? better to use normal rotation of sensor for new drivers

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    accumulate_sample(field, _instance, 20);
+}
+
+void AP_Compass_QMC5883P::read()
+{
+	GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "5883P Read ");
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:213]: remove debug

File: libraries/AP_Compass/AP_Compass_QMC5883P.h [hunk 12]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2016  Emlid Ltd. All rights reserved.
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_QMC5883P.h:2]: remove emlid copyright

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat [hunk 13]
```diff
@@ -127,9 +129,6 @@ define HAL_SUPPORT_RCOUT_SERIAL 1
...
 BARO MS56XX I2C:0:0x76
-COMPASS MMC5XX3 I2C:0:0x30 false ROTATION_NONE
-
-
-
-
-
+# COMPASS MMC5XX3 I2C:0:0x30 false ROTATION_NONE
+#COMPASS QMC5883L I2C:0:0x0D false ROTATION_ROLL_180_YAW_270
+COMPASS QMC5883P I2C:0:0x00 false ROTATION_ROLL_180_YAW_270
```
Comment (hendjoshsr71) [inline libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat:134]:
```suggestion
COMPASS QMC5883L I2C:0:0x0D false ROTATION_ROLL_180_YAW_270
```

File: libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat [hunk 14]
```diff
@@ -111,7 +113,7 @@ PA5 VSENSE2 ADC1 SCALE(1)
 PA6 VSENSE3 ADC1 SCALE(1)
 PB1 VSENSE4 ADC1 SCALE(1)
 
-define HAL_NO_GCS
+#define HAL_NO_GCS
```
Comment (hendjoshsr71) [inline libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat:116]: I think this can be removed completely? 
```suggestion
#define HAL_NO_GCS
```

File: libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat [hunk 15]
```diff
@@ -19,7 +19,8 @@ env AP_PERIPH 1
 # enable watchdog
 
 # crystal frequency
-OSCILLATOR_HZ 12000000
+# OSCILLATOR_HZ 12000000
+OSCILLATOR_HZ 0
```
Comment (hendjoshsr71) [inline libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat:23]: Why is this changed?

```suggestion
# OSCILLATOR_HZ 12000000
OSCILLATOR_HZ 0
```

File: libraries/AP_Compass/AP_Compass_QMC5883P.h [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <AP_Math/AP_Math.h>
+
+
+#include "AP_Compass.h"
+#include "AP_Compass_Backend.h"
+
+#ifndef HAL_COMPASS_QMC5883P_I2C_ADDR
+#define HAL_COMPASS_QMC5883P_I2C_ADDR 0x00
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_QMC5883P.h:29]: This is the correct address right?

```suggestion
#define HAL_COMPASS_QMC5883P_I2C_ADDR 0x0D
```

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+bool AP_Compass_QMC5883P::_check_whoami()
+{
+    uint8_t whoami;
+    //Affected by other devices,must read registers 0x00 once or reset,after can read the ID registers reliably
+    //    depreciated from L series to P
+    /**    _dev->read_registers(0x00,&whoami,1);
+    if (!_dev->read_registers(0x0C, &whoami,1)||
+      		whoami != 0x01){
+    	return false;
+    }**/
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:161]: Remove the dead code now that we know we dont need it.

```suggestion
    //Affected by other devices,must read registers 0x00 once or reset,after can read the ID registers reliably
    //    depreciated from L series to P
    /**    _dev->read_registers(0x00,&whoami,1);
    if (!_dev->read_registers(0x0C, &whoami,1)||
      		whoami != 0x01){
    	return false;
    }**/
```

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+#define QMC5883P_ODR_50HZ  (0x01 << 2)
+#define QMC5883P_ODR_100HZ (0x02 << 2)
+#define QMC5883P_ODR_200HZ (0x03 << 2)
+//
+// Over sampling Ratio OSR1
+#define QMC5883P_OSR1_8 (0x00 << 4)
+#define QMC5883P_OSR1_4 (0x01 << 4)
+#define QMC5883P_OSR1_2	(0x02 << 4)
+#define QMC5883P_OSR1_1	(0x03 << 4)
+//
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:56]:
```suggestion

```

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+#define QMC5883P_MODE_NORMAL     0x01
+#define QMC5883P_MODE_SINGLE     0x02
+#define QMC5883P_MODE_CONTINUOUS 0x03
+//
+// ODR data output rates for 5883L
+#define QMC5883P_ODR_10HZ  (0x00 << 2)
+#define QMC5883P_ODR_50HZ  (0x01 << 2)
+#define QMC5883P_ODR_100HZ (0x02 << 2)
+#define QMC5883P_ODR_200HZ (0x03 << 2)
+//
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:50]:
```suggestion

```

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+//Register Address
+#define QMC5883P_REG_ID                 0x00
+#define QMC5883P_REG_DATA_OUTPUT_X      0x01
+#define QMC5883P_REG_DATA_OUTPUT_Z_MSB  0x06
+#define QMC5883P_REG_STATUS             0x09
+#define QMC5883P_REG_CONF1              0x0A
+#define QMC5883P_REG_CONF2              0x0B
+
+#define QMC5883P_ID_VAL 0x80
+//
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp:37]:
```suggestion

```

File: libraries/AP_Compass/AP_Compass.cpp [hunk 21]
```diff
@@ -1181,7 +1200,8 @@ void Compass::_probe_external_i2c_compasses(void)
     const uint8_t rm3100_addresses[] = { HAL_COMPASS_RM3100_I2C_ADDR1,
                                          HAL_COMPASS_RM3100_I2C_ADDR2,
                                          HAL_COMPASS_RM3100_I2C_ADDR3,
-                                         HAL_COMPASS_RM3100_I2C_ADDR4 };
+                                         HAL_COMPASS_RM3100_I2C_ADDR4
+                                       };
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass.cpp:1204]: Undo all the whitespace fixes in the file and Compass.h

We try to keep the commit history only to functional changes. (Feel free to fix the whitespace if you the line though ;) )

```suggestion
                                         HAL_COMPASS_RM3100_I2C_ADDR4
                                       };
```

File: libraries/AP_Compass/AP_Compass.cpp [hunk 22]
```diff
@@ -1063,6 +1064,24 @@ void Compass::_probe_external_i2c_compasses(void)
...
+#if !defined(HAL_DISABLE_I2C_MAGS_BY_DEFAULT) || defined(HAL_USE_I2C_MAG_QMC5883P)
+    //external i2c bus
+    FOREACH_I2C_EXTERNAL(i) {
+        ADD_BACKEND(DRIVER_QMC5883P, AP_Compass_QMC5883P::probe(GET_I2C_DEVICE(i, HAL_COMPASS_QMC5883P_I2C_ADDR),
+                    true, HAL_COMPASS_QMC5883P_ORIENTATION_EXTERNAL));
+    }
+
+    // internal i2c bus
+    if (all_external) {
+        // only probe QMC5883L on internal if we are treating internals as externals
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass.cpp:1076]:
```suggestion
        // only probe QMC5883P on internal if we are treating internals as externals
```

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
Quick review since there are some formatting changes that need to be made.

Please don't create massive white space changes in files . (I know you just ran Astyle against all your modified files ;) )

You will need to squash your commits into three commits. One per library (feel to make the commit message more descriptive)

Tools: add add Compass QMC5883P
 hwdef: CarbonixL496 add Compass QMC5883P 
 AP_Compass: add QMC5883P

File: libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat (hunk 13 above)
Comment (Lokesh-Carbonix) [reply libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat]: changed it to 0x2C thanks

File: libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat (hunk 15 above)
Comment (Lokesh-Carbonix) [reply libraries/AP_HAL_ChibiOS/hwdef/CarbonixL496/hwdef.dat]: using internal oscillator so set to 0

File: libraries/AP_Compass/AP_Compass_QMC5883P.h (hunk 16 above)
Comment (Lokesh-Carbonix) [reply libraries/AP_Compass/AP_Compass_QMC5883P.h]: Set to correct address thanks.

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp (hunk 17 above)
Comment (Lokesh-Carbonix) [reply libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: thanks Removed the Dead code

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp (hunk 19 above)
Comment (Lokesh-Carbonix) [reply libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: Sorry its ok

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+ */
+#include "AP_Compass_QMC5883P.h"
+
+#include <stdio.h>
+#include <utility>
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <AP_Math/AP_Math.h>
+#include <GCS_MAVLink/GCS.h>
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: Dont need this now.
```suggestion
```

File: libraries/AP_Compass/AP_Compass_QMC5883P.cpp [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+ *
+ * Driver by Lokesh Ramina, Jan 2022
+ */
+#include "AP_Compass_QMC5883P.h"
+
+#include <stdio.h>
+#include <utility>
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/utility/sparse-endian.h>
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_QMC5883P.cpp]: Dont need this I think?

```suggestion
```

File: libraries/AP_Compass/AP_Compass_QMC5883P.h [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+/*
+  setup default orientations
+ */
+#ifndef HAL_COMPASS_QMC5883P_ORIENTATION_EXTERNAL
+#define HAL_COMPASS_QMC5883P_ORIENTATION_EXTERNAL ROTATION_ROLL_180
+#endif
+
+#ifndef HAL_COMPASS_QMC5883P_ORIENTATION_INTERNAL
+#define HAL_COMPASS_QMC5883P_ORIENTATION_INTERNAL ROTATION_ROLL_180_YAW_270
+#endif
```
Comment (hendjoshsr71) [inline libraries/AP_Compass/AP_Compass_QMC5883P.h:40]: Maybe these defaults should be ROTATION_NONE? I'm not sure. 

```suggestion
#ifndef HAL_COMPASS_QMC5883P_ORIENTATION_EXTERNAL
#define HAL_COMPASS_QMC5883P_ORIENTATION_EXTERNAL ROTATION_ROLL_180
#endif

#ifndef HAL_COMPASS_QMC5883P_ORIENTATION_INTERNAL
#define HAL_COMPASS_QMC5883P_ORIENTATION_INTERNAL ROTATION_ROLL_180_YAW_270
#endif
```

**Review by hendjoshsr71 — COMMENTED** [review]
LGTM once you test again with your board due to the number of changes and rebases :)

**Comment by andyp1per** [comment]
How different is this to the 5883L? If they are very similar we should consolidate the two drivers to save flash

**Comment by hendjoshsr71** [comment]
You are right Andy we should try to find a way to get this smaller.

On f103-periph, 
image

Adds 800B to Matek_F405 and 1.3KB to Durandal

The main differences are all the registers are shifted, different default address, some different configuration parameters, and some of the work arounds needed for the L are not necessary it seems.

File: libraries/AP_Compass/AP_Compass_QMC5883P.h (hunk 25 above)
Comment (Lokesh-Carbonix) [reply libraries/AP_Compass/AP_Compass_QMC5883P.h:40]: I am going to test it today as the board wasn't working over weekend.

**Comment by Lokesh-Carbonix** [comment]
The thing I am not sure about is if I mix both QMC5883L and P should we then change the naming of the file to something like QMC5883X and if yes? how will that affect the other code.

File: libraries/AP_Compass/AP_Compass_QMC5883P.h (hunk 25 above)
Comment (Lokesh-Carbonix) [reply libraries/AP_Compass/AP_Compass_QMC5883P.h:40]: > LGTM once you test again with your board due to the number of changes and rebases :)

The Compass is working properly after the code rebase.

**Comment by Lokesh-Carbonix** [comment]
@hendjoshsr71 @tridge should I add the functionality of QMC5883P into the QMC5883L library and keep the naming same or should I change the file name to QMC5883X? 

As changing the name will create a compatibility issue for people using QMC5883L, correct?  
and not changing the name will be confusing too.

**Comment by tridge** [comment]
> should I add the functionality of QMC5883P into the QMC5883L library and keep the naming same or should I change the file name to QMC5883X?

if you find that you can incorporate it into one drive in a reasonable fashion then yes, one driver is best. If you find it needs large changes then better as 2 drivers

renaming to QMC5883X is fine, it won't impact end users. Some hwdef.dat may need update, but that isn't a big deal and it will be caught by the compiler if someone is using it out of tree
